docs/rfcs/039-single-click-pg-upgrade.md

```diff
@@ -0,0 +1,199 @@
+**Need review from the storage team here** to understand how hard this restriction is and if there is any problem with +1.
+
+##### Where to set `S` for the new branch:
+
+- We need to use it as a start_lsn for the new Timeline struct
+- We need to set it in the control file of the new branch.
+- We need to set in in the Checkpoint record of the new branch.
+
+##### How to make the switch atomic:
+
+- We need to restrict writes to the old branch to get the `X` value and use it as `S` for the new branch.
+- IIUC, we need to do this right before the import. This means that timeline import time is customer visible pg_upgrade downtime. Can we somehow make it shorter? What is our target downtime for the upgrade? **Need storage team help here** 
+
+
+## Known technical issues:
+
+- We mantain relsize cache in neon. We need to copy it over to the new cluster during upgrade.
+
+- Fix import of the new version catalog into the new branch. We need to distinguish between catalog and non-catalog files and only skip the latter.
+Now this is done in a very hacky way, that will not work if any catalog files changed their relfilenode before the upgrade.
+**Need compute (postgres) team help here**
+
+- We need to update XIDs in the new version controlfile and skip SLRU files during timeline import. See copy_xact_xlog_xid() for details.
+
+- Proablby we need to pass some options to pg_upgrade. Same options that we use for initdb and start (i.e. locale)
+
+- We need to teach storage to handle the gap in the LSN space between the old branch end and new branch start for the scenario where `restore_lsn R` > `branch_lsn |`. See `~~~~~` on the picture, imagine case, where it is larger than `*****`.
+**Need storage team help here** to estimate the complexity of this task.
+
+- We need to run the "pg_upgrade microservice" somewhere.
+Possible options: 
+    - special mode of compute image (requires multi-version compute image).
+    - separate microservice running on pageserver machine. Synchronization with old running compute will be complicated here.
+
+## Open questions:
+
+**Need product, storage, compute and cplane teams help here**
+
+- Do we want to implement a full switch to the new version branch? Or do we want to start with "Test new version in a branch" feature? 
+Do we see product value in this feature?
+It is simpler to implenent, but we need to ensure that users won't use this new-version branch in production. How to do this?
+
+- Do we want to switch main endpoint to use upgraded branch? What if we upgraded from non-main branch?
+
+- How to restrict write access to the old branch after the siwtch?
+
+- How to communicate this pg_upgade switch to the user in UI / API?
+
+- How can we test the data correctness after the upgrade?
+
+- At what moment should we run recommended post-upgrade scripts?
+
+- How to handle extensions? This issue boils down to multi-version compute image.
```

Comment:
Interaction with `timeline_ancestor_detach`:

1. initially probably we want to just deny using this feature together with wal files from multiple postgres versions
    - or else we need to record the earlier pg version range which would be kind of extra
2. timeline ancestor detach in the presence of different pg versions will require rewriting all those pages which have not been modified in the branch instead of layer file copying
    - basically a reverse "image layer creation" operation we currently do with compaction
    - this might be already coming with archived branches or compaction enhancements